docs/spec/supervisor/supervisor.md

```diff
@@ -0,0 +1,153 @@
+
+```rust
+/// Stores the Light Blocks which could not be confirmed to have originated from
+/// the same chain as the predecesor light blocks.
+type FetchedStack = Stack<LightBlock>;
+
+/// Pair of light block and height of the light block which was highest 
+/// highest one for which we believe originated from the same chain
+type ChainPair = (LightBlock, Height);
+
+/// Traces the history of verify_to_target.
+trait Trace {
+    /// Returns the height of the current highest inserted
+    fn current_height(&self) -> Height;
+
+    /// Returns the current highest inserted light block.
+    fn current(&self) -> &LightBlock;
+
+    /// Insert a new highest light block. assert(self.current_height() < light_block.height())
+    fn insert(&mut self, light_block: LightBlock) -> Result<(), Error>;
+
+    /// Returns the list of all LightBlock which were inserted through 
+    /// insert with associated previous current_height().
+    fn verification_chain(&self) -> LinkedList<ChainPair>;
+}
+
+
+/// Given a primary and a starting_light_block confirm that the light block of target_height received
+/// from primary for target_height is from the same chain as the starting_light_block.
+/// Solution for forward skipping verification.
+pub fn verify_to_target(
+    primary: PeerID,
+    starting_light_block: LightBlock,
+    target_height: Height) -> Result<Trace, (LightBlock, Reason)> {
+
+    // Verified state always has one starting LightBlock
+    let mut trace = Trace::from(startingLightBlock);
+    let mut fetch_stack = FetchedStack::empty();
+
+    // insert the light block of target height, simplifies the loop
+    // we can say that at each iteration there should be something on stack
+    fetch_stack.push(fetch_light_block(primary, target_height));
```

Comment:
What does `fetch_light_block` do?

Reply:
It's just copied from the current specifications. In current code it's from the [`Io`](https://github.com/informalsystems/tendermint-rs/blob/5e298723f6f60ccf59b60c5bc7c202b338238260/light-client/src/components/io.rs#L71) component.

Reply:
It would greatly benefit if this side-effect is not done in this function, it could be a parameter to be called with.

Reply:
This was an error. The structure is changed a but now.